druntime/src/rt/lifetime.d

```diff
@@ -1754,89 +1754,81 @@ do
-mid-size arrays, making it very fast for medium arrays (for
-mid-to-large arrays, this turns out to be quite a bit faster than the
-equivalent realloc() code in C, on Linux at least. Small arrays are
-just as fast as GCC).
 - Perhaps most importantly, overall memory usage and stress on the GC
 is decreased significantly for demanding environments.
+- The algorithm is tuned to avoid any division at runtime.
 
 Params:
     newlength = new `.length`
-    size = old `.length`
+    elemsize = size of the element in the new array
 Returns: new capacity for array
 */
-size_t newCapacity(size_t newlength, size_t size)
+size_t newCapacity(size_t newlength, size_t elemsize)
 {
-    version (none)
-    {
-        size_t newcap = newlength * size;
-    }
-    else
-    {
-        size_t newcap = newlength * size;
-        size_t newext = 0;
+    size_t newcap = newlength * elemsize;
 
-        if (newcap > PAGESIZE)
+    /*
+     * Max growth factor numerator is 234, so allow for multiplying by 256.
+     * But also, the resulting size cannot be more than 2x, so prevent
+     * growing if 2x would fill up the address space (for 32-bit)
+     */
+    enum largestAllowed = (ulong.max >> 8) & (size_t.max >> 1);
+    if (newcap & ~largestAllowed)
+        return newcap;
+
+    /*
+     * The calculation for "extra" space depends on the requested capacity.
+     * We use an inverse logarithm of the new capacity to add an extra 15%
+     * to 83% capacity. Note that normally we humans think in terms of
+     * percent, but using 128 instead of 100 for the denominator means we
+     * can avoid all division by simply bit-shifthing. Since there are only
+     * 64 bits in a long, the bsr of a size_t is going to be 0 - 63. Using
+     * a lookup table allows us to precalculate the multiplier based on the
+     * inverse logarithm. The formula rougly is:
+     *
+     * newcap = request * (1.0 + min(0.83, 10.0 / (log(request) + 1)))
+     */
+    import core.bitop;
+    static immutable multTable = (){
+        assert(__ctfe);
+        ulong[size_t.sizeof * 8] result;
+        foreach (i; 0 .. result.length)
         {
-            //double mult2 = 1.0 + (size / log10(pow(newcap * 2.0,2.0)));
+            auto factor = 128 + 1280 / (i + 1);
+            result[i] = factor > 234 ? 234 : factor;
+        }
+        return result;
+    }();
 
-            // redo above line using only integer math
+    auto mult = multTable[bsr(newcap)];
 
-            /*static int log2plus1(size_t c)
-            {   int i;
+    // if this were per cent, then the code would look like:
+    // ((newlength * mult + 99) / 100) * elemsize
+    auto calccap = cast(size_t)((newlength * mult + 127) >> 7) * elemsize;
+    if(newcap > calccap)
+    {
+        import core.stdc.stdio;
+        static if(size_t.sizeof == 8)
+            auto slen = snprintf(errmsg.ptr, errmsg.length, "Oops! %lu %lu %lu %lu", calccap, newcap, newlength, elemsize, mult);
```

Comment:
Currently compilation fails on Windows because long is 32-bit there even on x64:
```
src\rt\lifetime.d(1823): Deprecation: argument `calccap` for format specification `"%lu"` must be `uint`, not `ulong`
src\rt\lifetime.d(1823):        C `long` is 4 bytes on your system
```

Reply:
Yes, I should have mentioned, the latest commit was meant to hopefully diagnose the possible issue. It actually did what I wanted it to do on x86 windows (built), but it's still hanging (fails after 40 minutes), so I didn't worry about this failure.

Without the latest commit, it compiles and passes tests on windows 64-bit.

I just need to try running locally and see why it's hanging.